### ISSUE TITLE
Composer wraps up with Fatal Error

### ISSUE THREAD

**Issue by rbroen** [laminas/laminas-skeleton-installer#14]
### Bug Report

|    Q        |   A
|------------ | ------
| laminas/laminas-mvc-skeleton  | fresh
| Ubuntu  | 20.10

#### Summary

Composer closes with a Fatal error and the Laminas installation questions are asked at every require action.

#### Current behavior

Composer closes with a Fatal error and the Laminas installation questions are asked at every require action.

#### How to reproduce

When running `composer create-project -s dev laminas/laminas-mvc-skeleton demo01` composer ends with the following error (but the additional packages questions were not asked):
```
Writing lock file
Generating autoload files
14 packages you are using are looking for funding.
Use the `composer fund` command to find out more!
PHP Fatal error:  Uncaught TypeError: Argument 1 passed to Symfony\Component\Console\Question\Question::__construct() must be of the type string, array given, called in /usr/share/php/Composer/IO/ConsoleIO.php on line 263 and defined in /usr/share/php/Symfony/Component/Console/Question/Question.php:38
Stack trace:
#0 /usr/share/php/Composer/IO/ConsoleIO.php(263): Symfony\Component\Console\Question\Question->__construct()
laminas/laminas-mvc-skeleton#1 /home/robert/repos/demo01/vendor/laminas/laminas-skeleton-installer/src/OptionalPackagesInstaller.php(151): Composer\IO\ConsoleIO->ask()
laminas/laminas-mvc-skeleton#2 /home/robert/repos/demo01/vendor/laminas/laminas-skeleton-installer/src/OptionalPackagesInstaller.php(83): Laminas\SkeletonInstaller\OptionalPackagesInstaller->requestMinimalInstall()
laminas/laminas-mvc-skeleton#3 /home/robert/repos/demo01/vendor/laminas/laminas-skeleton-installer/src/Plugin.php(84): Laminas\SkeletonInstaller\OptionalPackagesInstaller->__invoke()
laminas/laminas-mvc-skeleton#4 [internal function]: Laminas\SkeletonInstaller\Plugin->installOptionalDependencies()
laminas/tutorials#10 /usr/share/php/Composer/EventDispatcher/Even in /usr/share/php/Symfony/Component/Console/Question/Question.php on line 38

Fatal error: Uncaught TypeError: Argument 1 passed to Symfony\Component\Console\Question\Question::__construct() must be of the type string, array given, called in /usr/share/php/Composer/IO/ConsoleIO.php on line 263 and defined in /usr/share/php/Symfony/Component/Console/Question/Question.php:38
Stack trace:
#0 /usr/share/php/Composer/IO/ConsoleIO.php(263): Symfony\Component\Console\Question\Question->__construct()
laminas/laminas-mvc-skeleton#1 /home/robert/repos/demo01/vendor/laminas/laminas-skeleton-installer/src/OptionalPackagesInstaller.php(151): Composer\IO\ConsoleIO->ask()
laminas/laminas-mvc-skeleton#2 /home/robert/repos/demo01/vendor/laminas/laminas-skeleton-installer/src/OptionalPackagesInstaller.php(83): Laminas\SkeletonInstaller\OptionalPackagesInstaller->requestMinimalInstall()
laminas/laminas-mvc-skeleton#3 /home/robert/repos/demo01/vendor/laminas/laminas-skeleton-installer/src/Plugin.php(84): Laminas\SkeletonInstaller\OptionalPackagesInstaller->__invoke()
laminas/laminas-mvc-skeleton#4 [internal function]: Laminas\SkeletonInstaller\Plugin->installOptionalDependencies()
laminas/tutorials#10 /usr/share/php/Composer/EventDispatcher/Even in /usr/share/php/Symfony/Component/Console/Question/Question.php on line 38
```

The installation seems fine, bug the error will scare a lot of (new) users. And there is more...

When using the composer inside the docker container, the same errors appear after a composer require action (like: `composer require namshi/jose`), but first all extra package/module questions are asked and then the error occurs:
```
Updating root package
    Running an update to install optional packages

Fatal error: Uncaught ArgumentCountError: Too few arguments to function Composer\Installer::setUpdate(), 0 passed in /var/www/vendor/laminas/laminas-skeleton-installer/src/OptionalPackagesInstaller.php on line 321 and exactly 1 expected in phar:///usr/local/bin/composer/src/Composer/Installer.php:1053
Stack trace:
#0 /var/www/vendor/laminas/laminas-skeleton-installer/src/OptionalPackagesInstaller.php(321): Composer\Installer->setUpdate()
laminas/laminas-mvc-skeleton#1 /var/www/vendor/laminas/laminas-skeleton-installer/src/OptionalPackagesInstaller.php(107): Laminas\SkeletonInstaller\OptionalPackagesInstaller->runInstaller(Object(Composer\Package\RootPackage), Object(Laminas\SkeletonInstaller\Collection))
laminas/laminas-mvc-skeleton#2 phar:///usr/local/bin/composer/src/Composer/Plugin/PluginManager.php(213) : eval()'d code(84): Laminas\SkeletonInstaller\OptionalPackagesInstaller->__invoke()
laminas/laminas-mvc-skeleton#3 [internal function]: Laminas\SkeletonInstaller\Plugin_composer_tmp1->installOptionalDependencies(Object(Composer\Script\Event))
laminas/laminas-mvc-skeleton#4 phar:///usr/local/bin/composer/src/Composer/EventDispat in phar:///usr/local/bin/composer/src/Composer/Installer.php on line 1053
```

Every next require will result in the same actions/outcome.

#### Expected behavior

Fresh install of Laminas that gets new users into the mood of programming.

**Comment by rbroen** [comment]
I've done some digging....
On https://github.com/laminas/laminas-skeleton-installer/blob/0.3.x/src/OptionalPackagesInstaller.php line 321 `$installer->setUpdate();` is called without a parameter.
On https://github.com/composer/composer/blob/master/src/Composer/Installer.php on line 1053 `public function setUpdate($update)`, this does not seem an optional parameter....
I can imagine this being a problem, with or without strict mode.

**Comment by froschdesign** [comment]
@rbroen 
> I've done some digging....

Please add this comment to #13 – this related issue report for the problem with `Composer\Installer::setUpdate()`.

**Comment by froschdesign** [comment]
This error message has different problem:

 > ```bash
 > PHP Fatal error:
> Uncaught TypeError: Argument 1 passed to Symfony\Component\Console\Question\Question::__construct()
> must be of the type string, array given, called
> in /usr/share/php/Composer/IO/ConsoleIO.php on line 263 and defined
> in /usr/share/php/Symfony/Component/Console/Question/Question.php:38
>
> Stack trace:
> #0 /usr/share/php/Composer/IO/ConsoleIO.php(263): Symfony\Component\Console\Question\Question->__construct()
> #1 /home/robert/repos/demo01/vendor/laminas/laminas-skeleton-installer/src/OptionalPackagesInstaller.php(151): Composer\IO\ConsoleIO->ask()
> #2 /home/robert/repos/demo01/vendor/laminas/laminas-skeleton-installer/src/OptionalPackagesInstaller.php(83): Laminas\SkeletonInstaller\OptionalPackagesInstaller->requestMinimalInstall()
> #3 /home/robert/repos/demo01/vendor/laminas/laminas-skeleton-installer/src/Plugin.php(84): Laminas\SkeletonInstaller\OptionalPackagesInstaller->__invoke()
> #4 [internal function]: Laminas\SkeletonInstaller\Plugin->installOptionalDependencies()
> #5 /usr/share/php/Composer/EventDispatcher/Even in /usr/share/php/Symfony/Component/Console/Question/Question.php on line 38

During a _quick_ check I found the following: there is change of the `ask` method in `Composer\IO\IOInterface`.

In Composer version 1.6.3:

```php
    /**
     * Asks a question to the user.
     *
     * @param string|array $question The question to ask
     * @param string       $default  The default answer if none is given by the user
     *
     * @throws \RuntimeException If there is no data to read in the input stream
     * @return string            The user answer
     */
    public function ask($question, $default = null);
```
https://github.com/composer/composer/blob/88a69fda0f2187ad8714cedffd7a8872dceaa4c2/src/Composer/IO/IOInterface.php#L103-L112


In Composer version 1.6.4:

```php
    /**
     * Asks a question to the user.
     *
     * @param string $question The question to ask
     * @param string $default  The default answer if none is given by the user
     *
     * @throws \RuntimeException If there is no data to read in the input stream
     * @return string            The user answer
     */
    public function ask($question, $default = null);
```

https://github.com/composer/composer/blob/86ad51e8a3c64c9782446aae740a61fc6faa2522/src/Composer/IO/IOInterface.php#L103-L112

And since version 4.0.0 of Symfony's console component the `Question` class expected a string:

```php
    /**
     * @param string $question The question to ask to the user
     * @param mixed  $default  The default answer to return if the user enters nothing
     */
    public function __construct(string $question, $default = null)
```

https://github.com/symfony/console/blob/3179d1350843d2caa796973173dd0609ed4c7c67/Question/Question.php#L33-L37

The skeleton installer uses an `array` as parameter here:

https://github.com/laminas/laminas-skeleton-installer/blob/8beb996082ba17cde3bc04f347308019a89424c7/src/OptionalPackagesInstaller.php#L144-L151

---

_This is an inspection without IDE, so I can also be wrong here._

**Comment by weierophinney** [comment]
@froschdesign The problem arises from the fact that Composer v2 switched to symfony/console, and the interfaces and signatures changed. Compser v1 had its own IOInterface, and its `ask()` method allowed either a string or an array of strings as the first argument; v2 uses symfony/console's `Question` class instead, and the `ask()` method mandates a string for the first argument.

Fortunately, in auditing our usage of `ask()`, we always passed an array with a single string. I'm creating a patch now that passes that single value, instead of the array value.

**Comment by weierophinney** [comment]
Fixed with #16
